Feature Class definitions

msladek opened the pull request:
https://synjira.atlassian.net/browse/CELDEV-262

introduction of
- ClassField and implementations (StringField, DateField, etc.)
- ClassDefinition, defines a class via DocRef and ClassFields
- XClassCreator, uses ClassDefinitions to generate XWikis BaseClass

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,98 @@
+
+  private static Logger LOGGER = LoggerFactory.getLogger(DefaultXClassCreator.class);
+
+  static final String BLACKLIST_KEY = "celements.classdefinition.blacklist";
```
make constants private or public

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,98 @@
+  @Override
+  public void createXClasses() throws DocumentAccessException {
+    LOGGER.info("create classes for database '{}'", webUtils.getWikiRef());
+    for (ClassDefinition classDef : Utils.getComponentList(ClassDefinition.class)) {
```
why not defining a @Requirement for ClassDefinition list?

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,98 @@
+
+  private boolean isBlacklisted(ClassDefinition classDef) {
+    return configSrc.containsKey(BLACKLIST_KEY)
+        && configSrc.<List<?>>getProperty(BLACKLIST_KEY).contains(classDef.getName());
```
your looking up a String with contains(classDef.getName()). Thus you are expecting getProperty returns you a List<String>. Please add the appropriate generics.

fpichler commented on src/main/java/com/celements/common/classes/IClassesCompositorComponent.java:
```diff
@@ -5,6 +5,12 @@
 @ComponentRole
 public interface IClassesCompositorComponent {
 
+  public void checkClasses();
```
please add a javadoc explaining public or protected methods with a sentence or two

fpichler commented on src/main/java/com/celements/common/classes/XClassCreator.java:
```diff
@@ -0,0 +1,31 @@
+@ComponentRole
+public interface XClassCreator {
+
+  public void createXClasses() throws DocumentAccessException;
```
please add a javadoc explaining public or protected methods with a sentence or two

fpichler commented on src/main/java/com/celements/model/classes/ClassDefinition.java:
```diff
@@ -0,0 +1,21 @@
+@ComponentRole
+public interface ClassDefinition {
+
+  public String getName();
```
what is "getName" for. Please explain in javadoc

fpichler commented on src/main/java/com/celements/model/classes/fields/AbstractClassField.java:
```diff
@@ -0,0 +1,112 @@
+    return getWebUtils().serializeRef(getClassRef(), local) + "." + getName();
+  }
+
+  protected static IWebUtilsService getWebUtils() {
```
why not adding a @Requirement for IWebUtilsService?

msladek replied:
ClassFields are not designed as components because of (partial) immutability

fpichler replied:
ok for me

fpichler commented:
some issues and please add a CELDEV jira issue explaining or referencing the design.

msladek commented:
I fixed/addressed your suggestions, thanks. Also added [ListFields](https://github.com/celements/celements-core/pull/51/commits/95ffc1eeb2440e2b150845902df9b0cc6e248b5e), please review these too.

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,100 @@
+  private List<ClassDefinition> classDefinitions;
+
+  @Override
+  public void createXClasses() throws DocumentAccessException {
```
I think DocumentAccessException should be a RuntimeException (see related e-mail discussion)

Otherwise please throw a XClassCreationFailedException or similar

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,100 @@
+        BLACKLIST_KEY).contains(classDef.getName());
+  }
+
+  private void createXClass(ClassDefinition classDef) throws DocumentAccessException {
```
I think DocumentAccessException should be a RuntimeException (see related e-mail discussion)

Otherwise please throw a XClassCreationFailedException or similar.

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,100 @@
+      }
+    }
+    if (needsUpdate) {
+      modelAccess.saveDocument(classDoc);
```
please add a save comment indicating XClass was created or updated.

fpichler commented on src/main/java/com/celements/model/classes/ClassDefinition.java:
```diff
@@ -0,0 +1,33 @@
+  /**
+   * @return a list of all fields defining this class
+   */
+  public abstract List<ClassField<?>> getFields();
```
please do not use "abstract" in an interface. Interface is per definition "abstract"

fpichler commented on src/main/java/com/celements/model/classes/fields/AbstractClassField.java:
```diff
@@ -0,0 +1,112 @@
+import com.xpn.xwiki.objects.classes.PropertyClass;
+import com.xpn.xwiki.web.Utils;
+
+public abstract class AbstractClassField<T> implements ClassField<T> {
```
please add comment that this class is NOT thread-safe

msladek replied:
all lazy fields marked as volatile

fpichler commented:
Replacing the ClassCollection contract/feature the "collection" or package feature seams missing. Please see more specific explanation on https://synjira.atlassian.net/browse/CELDEV-262?focusedCommentId=17500&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17500

msladek commented:
first requires merge of #57 #58

fpichler commented:
please resolve conflicts

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,108 @@
+@Component
+public class DefaultXClassCreator implements XClassCreator {
+
+  private static Logger LOGGER = LoggerFactory.getLogger(DefaultXClassCreator.class);
```
please make final

fpichler commented on src/main/java/com/celements/common/classes/DefaultXClassCreator.java:
```diff
@@ -0,0 +1,108 @@
+  public void createXClasses(ClassPackage classPackage) throws XClassCreateException {
+    LOGGER.debug("creating package '{}'", classPackage.getName());
+    for (ClassDefinition classDef : classPackage.getClassDefinitions()) {
+      if (!classDef.isBlacklisted()) {
```
do we really need to fail completely if the creation of one class fails? Maybe we should try the next class or the next package...

fpichler commented on src/main/java/com/celements/model/classes/fields/AbstractClassField.java:
```diff
@@ -0,0 +1,117 @@
+import com.xpn.xwiki.web.Utils;
+
+/**
+ * Subclasses are expected be thread safe
```
grammer: ... expected to be thread safe

msladek commented:
edo please review this PR as well, these changes are worth a third pair of eyes :)

ebeutler commented:
looks fine as far as I could follow.

fpichler commented on src/main/java/com/celements/model/classes/fields/AbstractClassField.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * Subclasses are expected to be thread safe
+ */
+public abstract class AbstractClassField<T> implements ClassField<T> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/BooleanField.java:
```diff
@@ -0,0 +1,54 @@
+import com.xpn.xwiki.objects.classes.BooleanClass;
+import com.xpn.xwiki.objects.classes.PropertyClass;
+
+public class BooleanField extends AbstractClassField<Boolean> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/ClassField.java:
```diff
@@ -0,0 +1,26 @@
+
+import com.xpn.xwiki.objects.PropertyInterface;
+
+public interface ClassField<T> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/CustomClassField.java:
```diff
@@ -0,0 +1,13 @@
+
+import javax.annotation.Nullable;
+
+public interface CustomClassField<T> extends ClassField<T> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/DateField.java:
```diff
@@ -0,0 +1,69 @@
+import com.xpn.xwiki.objects.classes.DateClass;
+import com.xpn.xwiki.objects.classes.PropertyClass;
+
+public class DateField extends AbstractClassField<Date> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/StringField.java:
```diff
@@ -0,0 +1,41 @@
+import com.xpn.xwiki.objects.classes.PropertyClass;
+import com.xpn.xwiki.objects.classes.StringClass;
+
+public class StringField extends AbstractClassField<String> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/list/DBListField.java:
```diff
@@ -0,0 +1,36 @@
+import com.xpn.xwiki.objects.classes.DBListClass;
+import com.xpn.xwiki.objects.classes.ListClass;
+
+public class DBListField extends StringListField {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/list/EnumListField.java:
```diff
@@ -0,0 +1,59 @@
+import com.xpn.xwiki.objects.classes.ListClass;
+import com.xpn.xwiki.objects.classes.StaticListClass;
+
+public class EnumListField<T extends Enum<T>> extends ListField<T> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/list/ListField.java:
```diff
@@ -0,0 +1,122 @@
+import com.xpn.xwiki.objects.classes.ListClass;
+import com.xpn.xwiki.objects.classes.PropertyClass;
+
+public abstract class ListField<T> extends AbstractClassField<List<T>> implements
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/list/StaticListField.java:
```diff
@@ -0,0 +1,41 @@
+import com.xpn.xwiki.objects.classes.ListClass;
+import com.xpn.xwiki.objects.classes.StaticListClass;
+
+public class StaticListField extends StringListField {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/list/StringListField.java:
```diff
@@ -0,0 +1,31 @@
+
+import com.google.common.base.Joiner;
+
+public abstract class StringListField extends ListField<String> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/number/FloatField.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.xwiki.model.reference.DocumentReference;
+
+public class FloatField extends NumberField<Float> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/number/IntField.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.xwiki.model.reference.DocumentReference;
+
+public class IntField extends NumberField<Integer> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/number/LongField.java:
```diff
@@ -0,0 +1,18 @@
+
+import org.xwiki.model.reference.DocumentReference;
+
+public class LongField extends NumberField<Long> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/number/NumberField.java:
```diff
@@ -0,0 +1,38 @@
+import com.xpn.xwiki.objects.classes.NumberClass;
+import com.xpn.xwiki.objects.classes.PropertyClass;
+
+public abstract class NumberField<T extends Number> extends AbstractClassField<T> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/ref/AttachmentReferenceField.java:
```diff
@@ -0,0 +1,19 @@
+import org.xwiki.model.reference.AttachmentReference;
+import org.xwiki.model.reference.DocumentReference;
+
+public class AttachmentReferenceField extends EntityReferenceField<AttachmentReference> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/ref/EntityReferenceField.java:
```diff
@@ -0,0 +1,58 @@
+import com.xpn.xwiki.objects.classes.PropertyClass;
+import com.xpn.xwiki.objects.classes.StringClass;
+
+public abstract class EntityReferenceField<T extends EntityReference> extends AbstractClassField<T>
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/ref/SpaceReferenceField.java:
```diff
@@ -0,0 +1,19 @@
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.SpaceReference;
+
+public class SpaceReferenceField extends EntityReferenceField<SpaceReference> {
```
maybe add annotation @ThreadSafe

fpichler commented on src/main/java/com/celements/model/classes/fields/ref/WikiReferenceField.java:
```diff
@@ -0,0 +1,19 @@
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.WikiReference;
+
+public class WikiReferenceField extends EntityReferenceField<WikiReference> {
```
maybe add annotation @ThreadSafe

fpichler commented:
I think it should be possible to make ClassField implementations Immutable. This would simplify implementation significantly. I do not see the use case where this objects must be mutable.

msladek commented:
ClassFields are now immutable, please review.

fpichler commented on src/main/java/com/celements/common/classes/ClassesCompositorComponent.java:
```diff
@@ -81,13 +95,17 @@ void checkClassCollections() {
       }
     }
-    return false;
+    try {
```
please move "lookup" before the for loop

fpichler commented on src/main/java/com/celements/common/classes/XClassCreator.java:
```diff
@@ -0,0 +1,53 @@
+
+  /**
+   * creates XClasses from all {@link ClassPackage}s if activated, class definition isn't
+   * blacklisted and XClasses don't exist yet
```
if the XClass already exist, does it get updated (new fields added) or not? Thus maybe calling createAndUpdateXClasses is more precise?
